FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,80 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				if let existingCache = try ManagedCache.fetch(in: context) {
+					completion(.found(feed: existingCache.feed.toLocal(), timestamp: existingCache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let existingCache = try ManagedCache.fetch(in: context)
+				existingCache.map(context.delete)
+
+				let newCache = ManagedCache(context: context)
+				newCache.timestamp = timestamp
+				newCache.feed = feed.toManaged(context: context)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		context.perform { [context] in
+			do {
+				let existingCache = try ManagedCache.fetch(in: context)
+				existingCache.map(context.delete)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+}
+
+private extension NSOrderedSet {
+	func toLocal() -> [LocalFeedImage] {
+		array
+			.compactMap { $0 as? ManagedFeedImage }
```

Comment:
Since `NSOrderedSet` conforms to `Sequence`, it is possible to call `compactMap` directly, there's no need to convert it to an `array`. 
```suggestion
		compactMap { $0 as? ManagedFeedImage }
```